### PR TITLE
[SYCL] Enable USM address spaces generation under opt flag

### PR DESCRIPTION
**Pull request by MrSidims**
This patch serves two purposes:
1. It reverts shift of accessors to global buffers from __global to
   __global_device address space, which was considered as not a good
   idea until proven otherwise. Still this shift is enabled under
   -fsycl_enable_usm_address_spaces option for testing;
2. It fixes a mangling issue of functions that accepts arguments
   of multi_ptr<> type for non-SPIR targets (see:
   https://github.com/intel/llvm/issues/2039)

Plans for the future:
-fsycl_enable_usm_address_spaces option is a temporary option and
will be removed. To solve the mangling issue __ENABLE_USM_ADDR_SPACE__
macro will be added anyway under a combination of -fintelfpga and -Xs
flags. Future of accessor pointer at this point is unclear.

Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

### CODE REVIEW

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 1]
```diff
@@ -6183,6 +6183,13 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
     }
     if (Args.hasArg(options::OPT_fsycl_unnamed_lambda))
       CmdArgs.push_back("-fsycl-unnamed-lambda");
+
+    // Enable generation of USM address spaces as opt-in.
+    // __ENABLE_USM_ADDR_SPACE__ will be used during compilation of SYCL headers
+    if ((getToolChain().getTriple().getSubArch() ==
+             llvm::Triple::SPIRSubArch_fpga &&
+         Args.hasArg(options::OPT_fsycl_enable_usm_address_spaces)))
```
Comment (AGindinson) [inline clang/lib/Driver/ToolChains/Clang.cpp:6191]:
```suggestion
    if (getToolChain().getTriple().getSubArch() ==
            llvm::Triple::SPIRSubArch_fpga &&
        Args.hasArg(options::OPT_fsycl_enable_usm_address_spaces))
```

Comment (MrSidims) [reply clang/lib/Driver/ToolChains/Clang.cpp:6191]: Oops, I incidentally amended and forced-push. I applied this suggestion.

File: clang/test/Driver/sycl-offload.c [hunk 2]
```diff
@@ -668,7 +668,8 @@
 // CHK-TOOLS-GEN: clang-offload-wrapper{{.*}} "-o=[[OUTPUT5:.+\.bc]]" "-host=x86_64-unknown-linux-gnu" "-target=spir64_gen{{.*}}" "-kind=sycl" "[[OUTPUT4]]"
 // CHK-TOOLS-CPU: clang-offload-wrapper{{.*}} "-o=[[OUTPUT5:.+\.bc]]" "-host=x86_64-unknown-linux-gnu" "-target=spir64_x86_64{{.*}}" "-kind=sycl" "[[OUTPUT4]]"
 // CHK-TOOLS-AOT: llc{{.*}} "-filetype=obj" "-o" "[[OUTPUT6:.+\.o]]" "[[OUTPUT5]]"
-// CHK-TOOLS-FPGA: clang{{.*}} "-triple" "spir64_fpga-unknown-unknown-sycldevice" {{.*}} "-fsycl-int-header=[[INPUT1:.+\.h]]" "-faddrsig"
+// CHK-TOOLS-FPGA-USM-DISABLE: clang{{.*}} "-triple" "spir64_fpga-unknown-unknown-sycldevice" {{.*}} "-fsycl-int-header=[[INPUT1:.+\.h]]" "-faddrsig"
+// CHK-TOOLS-FPGA-USM-ENABLE: clang{{.*}} "-triple" "spir64_fpga-unknown-unknown-sycldevice" {{.*}} "-fsycl-int-header=[[INPUT1:.+\.h]]" "-D__ENABLE_USM_ADDR_SPACE__" "-faddrsig"
```
Comment (AGindinson) [inline clang/test/Driver/sycl-offload.c:672]: An optional "precision boost":
```suggestion
// CHK-TOOLS-FPGA-USM-DISABLE-NOT: clang{{.*}} "-triple" "spir64_fpga-unknown-unknown-sycldevice" {{.*}} "-fsycl-int-header=[[INPUT1:.+\.h]]" {{.*}}"-D__ENABLE_USM_ADDR_SPACE__"
// CHK-TOOLS-FPGA-USM-ENABLE: clang{{.*}} "-triple" "spir64_fpga-unknown-unknown-sycldevice" {{.*}} "-fsycl-int-header=[[INPUT1:.+\.h]]" {{.*}}"-D__ENABLE_USM_ADDR_SPACE__"
```

Comment (MrSidims) [reply clang/test/Driver/sycl-offload.c:672]: I don't think it's a "precision boost", since since it adds {{.*}} in a check string.

Comment (MrSidims) [reply clang/test/Driver/sycl-offload.c:672]: Also addition of "-NOT" makes INPUT1 variable undefined which, I suppose, will shoot later in checks.

**Pull request merged**